Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/html

cdumez opened the pull request:
#### 8b47c9cc90208f1859c2da053ad0d6c4c94953b6
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/html
<a href="https://bugs.webkit.org/show_bug.cgi?id=284092">https://bugs.webkit.org/show_bug.cgi?id=284092</a>

Reviewed by Darin Adler.

* Source/WebCore/dom/make_names.pl:
(printNamesHeaderFile):
(printNamesCppFile):
* Source/WebCore/html/FTPDirectoryDocument.cpp:
(WebCore::FTPDirectoryDocumentParser::FTPDirectoryDocumentParser):
(WebCore::wasLastDayOfMonth):
(WebCore::FTPDirectoryDocumentParser::append):
(WebCore::FTPDirectoryDocumentParser::finish):
* Source/WebCore/html/HTMLFontElement.cpp:
(WebCore::parseFontSize):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::preventsParentObjectFromExposure):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::convertEnumerationToString):

Canonical link: <a href="https://commits.webkit.org/287447@main">https://commits.webkit.org/287447@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/abf2dfc933ccb00bf15aa7f61147fe0bd6341d58

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79681 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58676 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84218 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30714 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81789 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67762 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6968 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62278 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20130 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82748 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52338 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72566 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42586 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49682 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26717 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29145 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70820 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27179 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85635 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6923 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4832 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70539 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7093 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68411 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69783 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13793 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12702 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12309 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6876 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6761 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10263 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8565 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/FTPDirectoryDocument.cpp:
```diff
@@ -67,14 +66,12 @@ class FTPDirectoryDocumentParser final : public HTMLDocumentParser {
+    void checkBuffer(size_t len = 10)
     {
-        if ((m_dest - m_buffer) > m_size - len) {
+        if (m_destIndex > (m_size - len)) {
```
Not sure it’s helpful to have added the parentheses around `m_size - len`.

darinadler commented on Source/WebCore/html/FTPDirectoryDocument.cpp:
```diff
@@ -67,14 +66,12 @@ class FTPDirectoryDocumentParser final : public HTMLDocumentParser {
-            int oldOffset = m_dest - m_buffer;
-            m_buffer = static_cast<UChar*>(fastRealloc(m_buffer, newSize * sizeof(UChar)));
-            m_dest = m_buffer + oldOffset;
+            m_buffer.realloc(newSize * sizeof(UChar));
```
Seems like this could overflow.

cdumez replied:
Used `CheckedSize` to guard against overflows.

darinadler commented on Source/WebCore/html/FTPDirectoryDocument.cpp:
```diff
@@ -67,14 +66,12 @@ class FTPDirectoryDocumentParser final : public HTMLDocumentParser {
-        if ((m_dest - m_buffer) > m_size - len) {
+        if (m_destIndex > (m_size - len)) {
             // Enlarge buffer
             int newSize = std::max(m_size * 2, m_size + len);
```
I wonder what guarantees this doesn’t overflow. Seems like this should just use something like `Vector<UChar>`; it’s such old code and so strange that it does its own allocation like this.

cdumez replied:
Ok, I switched to a Vector.

darinadler replied:
Since vectors have capacity and their own rules about allocating capacity as the vector gets larger, we can get rid of the logic about enlarging the buffer, and just resize it to the amount needed. This would eliminate all the `CheckedSize` stuff. I might come back and do that.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/html
https://bugs.webkit.org/show_bug.cgi?id=284092

Reviewed by Darin Adler.

* Source/WebCore/dom/make_names.pl:
(printNamesHeaderFile):
(printNamesCppFile):
* Source/WebCore/html/FTPDirectoryDocument.cpp:
(WebCore::FTPDirectoryDocumentParser::FTPDirectoryDocumentParser):
(WebCore::wasLastDayOfMonth):
(WebCore::FTPDirectoryDocumentParser::append):
(WebCore::FTPDirectoryDocumentParser::finish):
* Source/WebCore/html/HTMLFontElement.cpp:
(WebCore::parseFontSize):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::preventsParentObjectFromExposure):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::convertEnumerationToString):

Canonical link: https://commits.webkit.org/287447@main
